recommendation/pytorch/ncf.py

```diff
@@ -166,24 +134,134 @@ def main():
+    test_neg_items = [l[:, 1] for l in test_negatives]
+
+
+
+    #test_negs = test_negs.cuda()
+    # create items with real sample at last position
+    test_users = [l.reshape(-1,1).repeat(1,1+args.valid_negative) for l in test_users]
+    test_items = [torch.cat((a.reshape(-1,args.valid_negative), b), dim=1)
+            for a, b in zip(test_neg_items, test_pos)]
+    del test_ratings, test_neg_users, test_neg_items
+
+    # generate dup mask and real indice for exact same behavior on duplication compare to reference
+    # here we need a sort that is stable(keep order of duplicates)
+    # this is a version works on integer
+    sorted_items, indices = zip(*[torch.sort(l) for l in test_items]) # [1,1,1,2], [3,1,0,2]
+    sum_item_indices = [a.float()+b.float()/len(b[0]) 
+            for a, b in zip(sorted_items, indices)] #[1.75,1.25,1.0,2.5]
+    indices_order = [torch.sort(l)[1] for l in sum_item_indices] #[2,1,0,3]
+    stable_indices = [torch.gather(a, 1, b) 
+            for a, b in zip(indices, indices_order)] #[0,1,3,2]
+    # produce -1 mask
+    dup_mask = [(l[:,0:-1] == l[:,1:]) for l in sorted_items]
+    dup_mask = [torch.cat((torch.zeros_like(a, dtype=torch.uint8), b),dim=1)
+            for a, b in zip(test_pos, dup_mask)]
+    dup_mask = [torch.gather(a,1,b.sort()[1])
+            for a, b in zip(dup_mask, stable_indices)]
+    # produce real sample indices to later check in topk
+    sorted_items, indices = zip(*[(a != b).sort()
+            for a, b in zip(test_items, test_pos)])
+    sum_item_indices = [(a.float()) + (b.float())/len(b[0])
+            for a, b in zip(sorted_items, indices)]
+    indices_order = [torch.sort(l)[1] for l in sum_item_indices]
+    stable_indices = [torch.gather(a, 1, b)
+            for a, b in zip(indices, indices_order)]
+    real_indices = [l[:, 0] for l in stable_indices]
+    del sorted_items, indices, sum_item_indices, indices_order, stable_indices, test_pos
+
+    test_users = torch.cat(test_users)
+    test_items = torch.cat(test_items)
+    dup_mask = torch.cat(dup_mask)
+    real_indices = torch.cat(real_indices)
+
+    # make pytorch memory behavior more consistent later
+    torch.cuda.empty_cache()
 
     mlperf_log.ncf_print(key=mlperf_log.INPUT_BATCH_SIZE, value=args.batch_size)
-    mlperf_log.ncf_print(key=mlperf_log.INPUT_ORDER)  # set shuffle=True in DataLoader
-    train_dataloader = torch.utils.data.DataLoader(
-            dataset=train_dataset, batch_size=args.batch_size, shuffle=True,
-            num_workers=args.workers, pin_memory=True)
-    test_ratings = load_test_ratings(os.path.join(args.data, TEST_RATINGS_FILENAME))  # noqa: E501
-    test_negs = load_test_negs(os.path.join(args.data, TEST_NEG_FILENAME))
-    nb_users, nb_items = train_dataset.nb_users, train_dataset.nb_items
+    mlperf_log.ncf_print(key=mlperf_log.INPUT_ORDER)  # we shuffled later with randperm
+
     print('Load data done [%.1f s]. #user=%d, #item=%d, #train=%d, #test=%d'
```

Comment:
I find your code will first load all training data and test data, which requires larger memory, is it right? I run the code using a device with about 19G memory, it will say not have enough memory,  I double why not using pytorch dataloader to load this big dataset?

Reply:
Current version uses around 400GB of CPU memory on the x16x32 expanded dataset.

Pytorch dataloader is not fast enough for this use case because it concatenates the examples one-by-one.

Reply:
For the large dataset, it is important to set the --cpu_dataloader flag.  Trying to load the data on GPU memory is not feasible.

Reply:
Ok, I just have the machine with about 370G CPU memory, I will try it. I think it is not reasonable to load all data into memory, which costs much resources, by the way, if we use a bigger dataset than this one, how to do it? we should buy new memory card.